File: devel/libenkf/tests/enkf_fs.c
/*
   Copyright (C) 2013  Statoil ASA, Norway. 
    
   The file 'enkf_fs.c' is part of ERT - Ensemble based Reservoir Tool. 
    
   ERT is free software: you can redistribute it and/or modify 
   it under the terms of the GNU General Public License as published by 
   the Free Software Foundation, either version 3 of the License, or 
   (at your option) any later version. 
    
   ERT is distributed in the hope that it will be useful, but WITHOUT ANY 
   WARRANTY; without even the implied warranty of MERCHANTABILITY or 
   FITNESS FOR A PARTICULAR PURPOSE.   
    
   See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html> 
   for more details. 
*/


#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

#include <ert/util/test_util.h>
#include <ert/enkf/ert_test_context.h>
#include <ert/enkf/enkf_fs.h>
#include <ert/enkf/gen_kw.h>



void test_mount() {
  test_assert_false( enkf_fs_exists( "mnt" ));
  enkf_fs_create_fs("mnt" , BLOCK_FS_DRIVER_ID , NULL );
  bool read_only           = false;
  test_assert_true( enkf_fs_exists( "mnt" ));
  {
    enkf_fs_type * fs = enkf_fs_mount( "mnt" , read_only );
    test_assert_true( util_file_exists("mnt/mnt.lock"));
    test_assert_true( enkf_fs_is_instance( fs ));
    enkf_fs_decref( fs );
    test_assert_false( util_file_exists("mnt/mnt.lock"));
  }
}


void test_refcount() {
  enkf_fs_create_fs("mnt" , BLOCK_FS_DRIVER_ID , NULL );
  {
    enkf_fs_type * fs = enkf_fs_mount( "mnt" , false );
    test_assert_int_equal( 1 , enkf_fs_get_refcount( fs ));
    enkf_fs_decref( fs );
  }
}

void test_read_only() {
  enkf_fs_create_fs("mnt" , BLOCK_FS_DRIVER_ID , NULL );
  {
    {
      enkf_fs_type * fs_false = enkf_fs_mount( "mnt" , false );
      test_assert_false(enkf_fs_is_read_only(fs_false));
      test_assert_true( util_file_exists("mnt/mnt.lock"));
      enkf_fs_decref( fs_false );
    }

    {
      enkf_fs_type * fs_true = enkf_fs_mount( "mnt" , true );
      test_assert_true(enkf_fs_is_read_only(fs_true));
      test_assert_false( util_file_exists("mnt/mnt.lock"));
      enkf_fs_decref( fs_true );
    }
  }
}


void create_and_initalize_case(enkf_main_type * enkf_main, const char * new_case) {

  enkf_fs_type * initialized_case = enkf_main_mount_alt_fs(enkf_main, new_case, false, true);
  enkf_main_set_fs(enkf_main, initialized_case, NULL);
  test_assert_string_equal(enkf_main_get_current_fs(enkf_main), enkf_fs_get_case_name(initialized_case));

  int ens_size = enkf_main_get_ensemble_size(enkf_main);
  stringlist_type * param_list = stringlist_alloc_new();
  stringlist_append_copy(param_list, "MULTFLT");
  enkf_main_initialize_from_scratch(enkf_main, param_list, 0, ens_size-1, INIT_FORCE);

  enkf_fs_decref(initialized_case);
}


void test_invalidate_cache(ert_test_context_type * test_context) {
  enkf_main_type * enkf_main = ert_test_context_get_main(test_context);
  test_assert_not_NULL(enkf_main);

  const char * case1 = "test_case1";
  const char * case2 = "test_case2";
  create_and_initalize_case(enkf_main, case1);
  create_and_initalize_case(enkf_main, case2);

  int ensemble_size = enkf_main_get_ensemble_size(enkf_main);

  for (int iens = 0; iens < ensemble_size; ++iens) {

    double val1;
    double val2;

    node_id_type node_id = {.report_step = 0 ,
                            .iens = iens,
                            .state = ANALYZED };

    {
      enkf_main_select_fs(enkf_main, case1);
      test_assert_string_equal(enkf_main_get_current_fs(enkf_main),  case1);

      enkf_state_type * enkf_state = enkf_main_iget_state(enkf_main, iens);
      enkf_node_type * gen_kw_node = enkf_state_get_node(enkf_state, "MULTFLT");
      test_assert_not_NULL(gen_kw_node);

      test_assert_true(enkf_node_try_load(gen_kw_node, enkf_main_get_fs(enkf_main), node_id));
      gen_kw_type * kw_case1 = enkf_node_value_ptr(gen_kw_node);
      val1 = gen_kw_data_iget( kw_case1 , 0 , false );
    }

    {
      enkf_main_select_fs(enkf_main, case2);
      test_assert_string_equal(enkf_main_get_current_fs(enkf_main),  case2);

      enkf_state_type * enkf_state = enkf_main_iget_state(enkf_main, iens);
      enkf_node_type * gen_kw_node = enkf_state_get_node(enkf_state, "MULTFLT");
      test_assert_not_NULL(gen_kw_node);

      test_assert_true(enkf_node_try_load(gen_kw_node, enkf_main_get_fs(enkf_main), node_id));
      gen_kw_type * kw_case2 = enkf_node_value_ptr(gen_kw_node);
      val2 = gen_kw_data_iget( kw_case2, 0 , false );
    }

    test_assert_double_not_equal(val1, val2);
    printf("Realization nr %d: MULTFLT value from test_case 1 is %f, MULTFLT value from test_case2 is %f\n", iens+1, val1, val2);
  }
}



int main(int argc, char ** argv) {
  const char * config_file = argv[1];
  test_assert_not_NULL(config_file);
  ert_test_context_type * test_context = ert_test_context_alloc("test_context_enkf_fs_test" , config_file , NULL);

  test_mount();
  test_refcount();
  test_read_only();
  test_invalidate_cache(test_context);

  ert_test_context_free(test_context);
  exit(0);
}

File: src/res/enkf/enkf_fs.py
#  Copyright (C) 2012  Equinor ASA, Norway.
#
#  The file 'enkf_fs.py' is part of ERT - Ensemble based Reservoir Tool.
#
#  ERT is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ERT is distributed in the hope that it will be useful, but WITHOUT ANY
#  WARRANTY; without even the implied warranty of MERCHANTABILITY or
#  FITNESS FOR A PARTICULAR PURPOSE.
#
#  See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
#  for more details.
from typing import List

from cwrap import BaseCClass
import numpy as np
import numpy.typing as npt

from res import ResPrototype
from res.enkf.res_config import EnsembleConfig
from res.enkf.enums import EnKFFSType
from res.enkf.state_map import StateMap
from res.enkf.summary_key_set import SummaryKeySet
from res.enkf.util import TimeMap

from res._lib import update


class EnkfFs(BaseCClass):
    TYPE_NAME = "enkf_fs"

    _mount = ResPrototype("void* enkf_fs_mount(char*, bool)", bind=False)
    _sync = ResPrototype("void enkf_fs_sync(enkf_fs)")
    _decref = ResPrototype("int   enkf_fs_decref(enkf_fs)")
    _incref = ResPrototype("int   enkf_fs_incref(enkf_fs)")
    _get_refcount = ResPrototype("int   enkf_fs_get_refcount(enkf_fs)")
    _get_case_name = ResPrototype("char* enkf_fs_get_case_name(enkf_fs)")
    _is_read_only = ResPrototype("bool  enkf_fs_is_read_only(enkf_fs)")
    _is_running = ResPrototype("bool  enkf_fs_is_running(enkf_fs)")
    _fsync = ResPrototype("void  enkf_fs_fsync(enkf_fs)")
    _create = ResPrototype(
        "enkf_fs_obj   enkf_fs_create_fs(char* , enkf_fs_type_enum , bool)",
        bind=False,
    )
    _get_time_map = ResPrototype("time_map_ref  enkf_fs_get_time_map(enkf_fs)")
    _get_state_map = ResPrototype("state_map_ref enkf_fs_get_state_map(enkf_fs)")
    _summary_key_set = ResPrototype(
        "summary_key_set_ref enkf_fs_get_summary_key_set(enkf_fs)"
    )

    def __init__(self, mount_point: str, read_only: bool = False):
        c_ptr = self._mount(mount_point, read_only)
        super().__init__(c_ptr)

    def copy(self):
        fs = self.createPythonObject(self._address())
        self._incref()
        return fs

    # This method will return a new Python object which shares the underlying
    # enkf_fs instance as self. The name weakref is used because the Python
    # object returned from this method does *not* manipulate the reference
    # count of the underlying enkf_fs instance, and specifically it does not
    # inhibit destruction of this object.
    def weakref(self):
        fs = self.createCReference(self._address())
        return fs

    def getTimeMap(self) -> TimeMap:
        return self._get_time_map().setParent(self)

    def getStateMap(self) -> StateMap:
        return self._get_state_map().setParent(self)

    def getCaseName(self) -> str:
        return self._get_case_name()

    def isReadOnly(self) -> bool:
        return self._is_read_only()

    def refCount(self) -> int:
        return self._get_refcount()

    def is_running(self) -> bool:
        return self._is_running()

    @classmethod
    def createFileSystem(cls, path: str, mount: bool = False) -> "EnkfFs":
        assert isinstance(path, str)
        fs_type = EnKFFSType.BLOCK_FS_DRIVER_ID
        fs = cls._create(path, fs_type, mount)
        return fs

    def sync(self):
        self._sync()

    # The umount( ) method should not normally be called explicitly by
    # downstream code, but in situations where file descriptors is at premium
    # it might be beneficial to call it explicitly. In that case it is solely
    # the responsability of the calling scope to ensure that it is not called
    # repeatedly - that will lead to hard failure!
    def umount(self):
        if self.isReference():
            raise AssertionError(
                "Calling umount() on a reference is an application error"
            )

        if self:
            self._decref()
            self._invalidateCPointer()
        else:
            raise AssertionError("Tried to umount for second time - application error")

    def free(self):
        if self:
            self.umount()

    def __repr__(self):
        return f"EnkfFs(case_name = {self.getCaseName()}) {self._ad_str()}"

    def fsync(self):
        self._fsync()

    def getSummaryKeySet(self) -> SummaryKeySet:
        return self._summary_key_set().setParent(self)

    def realizationList(self, state):
        """
        Will return list of realizations with state == the specified state.
        @type state: res.enkf.enums.RealizationStateEnum
        @rtype: ecl.util.IntVector
        """
        state_map = self.getStateMap()
        return state_map.realizationList(state)

    def save_parameters(
        self,
        ensemble_config: EnsembleConfig,
        iens_active_index: List[int],
        parameters: List[update.Parameter],
        values: npt.ArrayLike,
    ):
        update.save_parameters(
            self, ensemble_config, iens_active_index, parameters, values
        )

    def load_parameters(
        self,
        ensemble_config: EnsembleConfig,
        iens_active_index: List[int],
        parameters: List[update.Parameter],
    ) -> np.ndarray:
        return update.load_parameters(
            self, ensemble_config, iens_active_index, parameters
        )
